src/errors.rs

```diff
@@ -1,50 +1,64 @@
-use super::Box;
+use super::{dealloc, Channel};
 use core::fmt;
 use core::mem;
+use core::ptr::NonNull;
 
 /// An error returned when trying to send on a closed channel. Returned from
 /// [`Sender::send`] if the corresponding [`Receiver`] has already been dropped.
 ///
 /// The message that could not be sent can be retreived again with [`SendError::into_inner`].
 pub struct SendError<T> {
-    channel_ptr: *mut super::Channel<T>,
+    channel_ptr: NonNull<Channel<T>>,
```

Comment:
Since it should preferably be cheap to sent and ignore the error. Not having to pay for that if you don't use the error for anything

Reply:
That makes sense. One thing to note is that the *correct* way to fully implement box involves using the unstable `dropck_eyepatch` feature to add `#[may_dangle]` to the `Drop` implementation, but all that does it make it slightly *less* restrictive which is better than the alternative.

TL;DR we should add a `PhantomData<T>` field, which I'll do later today.

Reply:
Great! To add to my last point: SendError should be small, so that the stack size to call Sender::send stays small (for small T).

Reply:
Yeah that's a good point as well, and `NonNull` helps with that too since `Result<(), SendError>` will decrease from 16 bytes to 8 (on 64-bit architectures), so it can fit in a register.

Reply:
This is a great improvement! Consider adding this as a test, so we make sure the null-pointer optimization does not have a regression further down the line. We can add this to the existing `error_sized` test:

```rust
    // Makes sure that the return value of `Sender::send` stays small. In essence
    // checks that the null-pointer optimization in `SendError` works.
    assert_eq!(
        mem::size_of::<Result<(), oneshot::SendError<[u8; 1024]>>>(),
        EXPECTED
    );
```

Reply:
Ah yeah I did add a test for that in one of the later commits from the draft PR. I'll move those over and deal with the rebase conflicts later cause it's a small easy change.